File: app/src/main/java/com/zakeercareer/calculator/util/UnitConverter.kt
package com.zakeercareer.calculator.util

enum class UnitCategory(val displayName: String) {
    LENGTH("Length"),
    MASS("Mass"),
    TEMPERATURE("Temperature"),
    AREA("Area"),
    VOLUME("Volume"),
    SPEED("Speed"),
    TIME("Time"),
    PRESSURE("Pressure"),
    ENERGY("Energy"),
    DATA("Data Storage")
}

data class UnitItem(val name: String, val symbol: String, val factorToBase: Double)

object UnitConverter {

    fun getUnits(category: UnitCategory): List<UnitItem> {
        return when (category) {
            UnitCategory.LENGTH -> listOf(
                UnitItem("Meter", "m", 1.0),
                UnitItem("Kilometer", "km", 1000.0),
                UnitItem("Centimeter", "cm", 0.01),
                UnitItem("Millimeter", "mm", 0.001),
                UnitItem("Mile", "mi", 1609.344),
                UnitItem("Yard", "yd", 0.9144),
                UnitItem("Foot", "ft", 0.3048),
                UnitItem("Inch", "in", 0.0254)
            )
            UnitCategory.MASS -> listOf(
                UnitItem("Kilogram", "kg", 1.0),
                UnitItem("Gram", "g", 0.001),
                UnitItem("Milligram", "mg", 0.000001),
                UnitItem("Pound", "lb", 0.45359237),
                UnitItem("Ounce", "oz", 0.028349523125),
                UnitItem("Metric Ton", "t", 1000.0)
            )
            UnitCategory.TEMPERATURE -> listOf(
                UnitItem("Celsius", "°C", 1.0),
                UnitItem("Fahrenheit", "°F", 1.0),
                UnitItem("Kelvin", "K", 1.0)
            )
            UnitCategory.AREA -> listOf(
                UnitItem("Square Meter", "m²", 1.0),
                UnitItem("Square Kilometer", "km²", 1000000.0),
                UnitItem("Square Foot", "ft²", 0.09290304),
                UnitItem("Acre", "ac", 4046.8564224),
                UnitItem("Hectare", "ha", 10000.0)
            )
            UnitCategory.VOLUME -> listOf(
                UnitItem("Liter", "L", 1.0),
                UnitItem("Milliliter", "mL", 0.001),
                UnitItem("Cubic Meter", "m³", 1000.0),
                UnitItem("US Gallon", "gal", 3.785411784),
                UnitItem("US Cup", "cup", 0.2365882365),
                UnitItem("Fluid Ounce", "fl oz", 0.0295735295625)
            )
            UnitCategory.SPEED -> listOf(
                UnitItem("Meter / second", "m/s", 1.0),
                UnitItem("Kilometer / hour", "km/h", 1.0 / 3.6),
                UnitItem("Miles / hour", "mph", 0.44704),
                UnitItem("Knot", "kt", 1852.0 / 3600.0)
            )
            UnitCategory.TIME -> listOf(
                UnitItem("Second", "s", 1.0),
                UnitItem("Minute", "min", 60.0),
                UnitItem("Hour", "h", 3600.0),
                UnitItem("Day", "d", 86400.0),
                UnitItem("Week", "wk", 604800.0)
            )
            UnitCategory.PRESSURE -> listOf(
                UnitItem("Pascal", "Pa", 1.0),
                UnitItem("Bar", "bar", 100000.0),
                UnitItem("Atmosphere", "atm", 101325.0),
                UnitItem("PSI", "psi", 6894.757293168)
            )
            UnitCategory.ENERGY -> listOf(
                UnitItem("Joule", "J", 1.0),
                UnitItem("Kilojoule", "kJ", 1000.0),
                UnitItem("Calorie", "cal", 4.184),
                UnitItem("Kilocalorie", "kcal", 4184.0),
                UnitItem("Watt-hour", "Wh", 3600.0)
            )
            UnitCategory.DATA -> listOf(
                UnitItem("Byte", "B", 1.0),
                UnitItem("Kilobyte", "KB", 1024.0),
                UnitItem("Megabyte", "MB", 1048576.0),
                UnitItem("Gigabyte", "GB", 1073741824.0),
                UnitItem("Terabyte", "TB", 1099511627776.0)
            )
        }
    }

    fun convert(
        value: Double,
        category: UnitCategory,
        fromUnit: UnitItem,
        toUnit: UnitItem
    ): Double {
        if (category == UnitCategory.TEMPERATURE) {
            return convertTemperature(value, fromUnit.symbol, toUnit.symbol)
        }
        val baseValue = value * fromUnit.factorToBase
        return baseValue / toUnit.factorToBase
    }

    private fun convertTemperature(value: Double, fromSymbol: String, toSymbol: String): Double {
        if (fromSymbol == toSymbol) return value
        val celsius = when (fromSymbol) {
            "°C" -> value
            "°F" -> (value - 32.0) * 5.0 / 9.0
            "K" -> value - 273.15
            else -> value
        }
        return when (toSymbol) {
            "°C" -> celsius
            "°F" -> (celsius * 9.0 / 5.0) + 32.0
            "K" -> celsius + 273.15
            else -> celsius
        }
    }
}

File: src/utils/unitConverter.ts
import { UnitCategory, UnitItem } from '../types';

export const UNIT_CATEGORIES: UnitCategory[] = [
  'Length',
  'Mass',
  'Temperature',
  'Area',
  'Volume',
  'Speed',
  'Time',
  'Pressure',
  'Energy',
  'Data Storage',
];

export const UNIT_DATA: Record<UnitCategory, UnitItem[]> = {
  Length: [
    { name: 'Meter', symbol: 'm', factorToBase: 1.0 },
    { name: 'Kilometer', symbol: 'km', factorToBase: 1000.0 },
    { name: 'Centimeter', symbol: 'cm', factorToBase: 0.01 },
    { name: 'Millimeter', symbol: 'mm', factorToBase: 0.001 },
    { name: 'Mile', symbol: 'mi', factorToBase: 1609.344 },
    { name: 'Yard', symbol: 'yd', factorToBase: 0.9144 },
    { name: 'Foot', symbol: 'ft', factorToBase: 0.3048 },
    { name: 'Inch', symbol: 'in', factorToBase: 0.0254 },
  ],
  Mass: [
    { name: 'Kilogram', symbol: 'kg', factorToBase: 1.0 },
    { name: 'Gram', symbol: 'g', factorToBase: 0.001 },
    { name: 'Milligram', symbol: 'mg', factorToBase: 0.000001 },
    { name: 'Pound', symbol: 'lb', factorToBase: 0.45359237 },
    { name: 'Ounce', symbol: 'oz', factorToBase: 0.028349523125 },
    { name: 'Metric Ton', symbol: 't', factorToBase: 1000.0 },
  ],
  Temperature: [
    { name: 'Celsius', symbol: '°C', factorToBase: 1.0 },
    { name: 'Fahrenheit', symbol: '°F', factorToBase: 1.0 },
    { name: 'Kelvin', symbol: 'K', factorToBase: 1.0 },
  ],
  Area: [
    { name: 'Square Meter', symbol: 'm²', factorToBase: 1.0 },
    { name: 'Square Kilometer', symbol: 'km²', factorToBase: 1000000.0 },
    { name: 'Square Foot', symbol: 'ft²', factorToBase: 0.09290304 },
    { name: 'Acre', symbol: 'ac', factorToBase: 4046.8564224 },
    { name: 'Hectare', symbol: 'ha', factorToBase: 10000.0 },
  ],
  Volume: [
    { name: 'Liter', symbol: 'L', factorToBase: 1.0 },
    { name: 'Milliliter', symbol: 'mL', factorToBase: 0.001 },
    { name: 'Cubic Meter', symbol: 'm³', factorToBase: 1000.0 },
    { name: 'US Gallon', symbol: 'gal', factorToBase: 3.785411784 },
    { name: 'US Cup', symbol: 'cup', factorToBase: 0.2365882365 },
    { name: 'Fluid Ounce', symbol: 'fl oz', factorToBase: 0.0295735295625 },
  ],
  Speed: [
    { name: 'Meter / second', symbol: 'm/s', factorToBase: 1.0 },
    { name: 'Kilometer / hour', symbol: 'km/h', factorToBase: 0.277777778 },
    { name: 'Miles / hour', symbol: 'mph', factorToBase: 0.44704 },
    { name: 'Knot', symbol: 'kt', factorToBase: 0.514444 },
  ],
  Time: [
    { name: 'Second', symbol: 's', factorToBase: 1.0 },
    { name: 'Minute', symbol: 'min', factorToBase: 60.0 },
    { name: 'Hour', symbol: 'h', factorToBase: 3600.0 },
    { name: 'Day', symbol: 'd', factorToBase: 86400.0 },
    { name: 'Week', symbol: 'wk', factorToBase: 604800.0 },
  ],
  Pressure: [
    { name: 'Pascal', symbol: 'Pa', factorToBase: 1.0 },
    { name: 'Bar', symbol: 'bar', factorToBase: 100000.0 },
    { name: 'Atmosphere', symbol: 'atm', factorToBase: 101325.0 },
    { name: 'PSI', symbol: 'psi', factorToBase: 6894.757293168 },
  ],
  Energy: [
    { name: 'Joule', symbol: 'J', factorToBase: 1.0 },
    { name: 'Kilojoule', symbol: 'kJ', factorToBase: 1000.0 },
    { name: 'Kilocalorie', symbol: 'kcal', factorToBase: 4184.0 },
    { name: 'Watt-hour', symbol: 'Wh', factorToBase: 3600.0 },
    { name: 'Kilowatt-hour', symbol: 'kWh', factorToBase: 3600000.0 },
  ],
  'Data Storage': [
    { name: 'Byte', symbol: 'B', factorToBase: 1.0 },
    { name: 'Kilobyte', symbol: 'KB', factorToBase: 1024.0 },
    { name: 'Megabyte', symbol: 'MB', factorToBase: 1048576.0 },
    { name: 'Gigabyte', symbol: 'GB', factorToBase: 1073741824.0 },
    { name: 'Terabyte', symbol: 'TB', factorToBase: 1099511627776.0 },
  ],
};

export function convertUnits(
  category: UnitCategory,
  fromUnit: UnitItem,
  toUnit: UnitItem,
  value: number
): number {
  if (isNaN(value)) return 0;
  if (fromUnit.name === toUnit.name) return value;

  if (category === 'Temperature') {
    // Convert from -> Celsius
    let celsius = value;
    if (fromUnit.name === 'Fahrenheit') {
      celsius = ((value - 32) * 5) / 9;
    } else if (fromUnit.name === 'Kelvin') {
      celsius = value - 273.15;
    }

    // Convert Celsius -> to
    if (toUnit.name === 'Celsius') return celsius;
    if (toUnit.name === 'Fahrenheit') return (celsius * 9) / 5 + 32;
    if (toUnit.name === 'Kelvin') return celsius + 273.15;
    return celsius;
  }

  // Base unit conversion: value * fromFactor / toFactor
  const baseValue = value * fromUnit.factorToBase;
  return baseValue / toUnit.factorToBase;
}
